### ISSUE TITLE
Regression in 1.20.0-dev with Modelica.Mechanics.MultiBody.Examples.Loops.PlanarLoops_analytic

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8957]
### Description

Modelica.Mechanics.MultiBody.Examples.Loops.PlanarLoops_analytic worked in 1.19.0-beta.1. In the current 1.20.0-dev nightly, it breaks during C compilation with

```
PlanarLoops_analytic_02nls.c:147:58: error: invalid operands to binary expression ('real_array' (aka 'struct base_array_s') and 'double')
  real_array_copy_data(sub_alloc_real_array(tmp0, (tmp1) * (sqrt(tmp4))), tmp5);
                                                  ~~~~~~ ^ ~~~~~~~~~~~~
C:/Program Files/OpenModelica1.20.0-dev-64bit/include/omc/c/util/generic_array.h:48:76: note: expanded from macro 'real_array_copy_data'
#define real_array_copy_data(src,dst)               simple_array_copy_data(src, &dst, sizeof(modelica_real));
                                                                           ^~~
PlanarLoops_analytic_02nls.c:411:58: error: invalid operands to binary expression ('real_array' (aka 'struct base_array_s') and 'double')
  real_array_copy_data(sub_alloc_real_array(tmp0, (tmp1) * (sqrt(tmp4))), tmp5);
                                                  ~~~~~~ ^ ~~~~~~~~~~~~
C:/Program Files/OpenModelica1.20.0-dev-64bit/include/omc/c/util/generic_array.h:48:76: note: expanded from macro 'real_array_copy_data'
#define real_array_copy_data(src,dst)               simple_array_copy_data(src, &dst, sizeof(modelica_real));
                                                                           ^~~
PlanarLoops_analytic_02nls.c:675:58: error: invalid operands to binary expression ('real_array' (aka 'struct base_array_s') and 'double')
  real_array_copy_data(sub_alloc_real_array(tmp0, (tmp1) * (sqrt(tmp4))), tmp5);
                                                  ~~~~~~ ^ ~~~~~~~~~~~~
C:/Program Files/OpenModelica1.20.0-dev-64bit/include/omc/c/util/generic_array.h:48:76: note: expanded from macro 'real_array_copy_data'
#define real_array_copy_data(src,dst)               simple_array_copy_data(src, &dst, sizeof(modelica_real));
                                                                           ^~~
clang  -Os -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse     -I"C:/Program Files/OpenModelica1.20.0-dev-64bit/include/omc/c" -I"C:/Program Files/OpenModelica1.20.0-dev-64bit/include/omc" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=PlanarLoops_analytic -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=5 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o PlanarLoops_analytic_06inz.o PlanarLoops_analytic_06inz.c
3 errors generated.
mingw32-make: *** [<builtin>: PlanarLoops_analytic_02nls.o] Error 1
mingw32-make: *** Waiting for unfinished jobs....
PlanarLoops_analytic_06inz.c:11958:59: error: assigning to 'modelica_real' (aka 'double') from incompatible type 'real_array_t' (aka 'struct base_array_s')
  data->simulationInfo->nonlinearSystemData[0].nlsxOld[0] = division_alloc_real_array_scalar(threadData,tmp0,sqrt(tmp3),"sqrt(jointRRR1.jointUSR.revolute.n * jointRRR1.jointUSR.revolute.n)");
                                                          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
PlanarLoops_analytic_06inz.c:12151:59: error: assigning to 'modelica_real' (aka 'double') from incompatible type 'real_array_t' (aka 'struct base_array_s')
  data->simulationInfo->nonlinearSystemData[1].nlsxOld[0] = division_alloc_real_array_scalar(threadData,tmp0,sqrt(tmp3),"sqrt(jointRRR2.jointUSR.revolute.n * jointRRR2.jointUSR.revolute.n)");
                                                          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
PlanarLoops_analytic_06inz.c:12389:59: error: assigning to 'modelica_real' (aka 'double') from incompatible type 'real_array_t' (aka 'struct base_array_s')
  data->simulationInfo->nonlinearSystemData[2].nlsxOld[0] = division_alloc_real_array_scalar(threadData,tmp0,sqrt(tmp3),"sqrt(jointRRR3.jointUSR.revolute.n * jointRRR3.jointUSR.revolute.n)");
                                                          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
mingw32-make: *** [<builtin>: PlanarLoops_analytic_06inz.o] Error 1
mingw32-make: Leaving directory 'C:/Users/FRANCE~1/AppData/Local/Temp/OPENMO~1/OMEdit/MODELI~1.PLA'
Compilation process failed. Exited with code 0x00000002.
```
Unfortunately the library testsuite has been broken for a while, so it won't be trival to figure out the root cause.

**Comment by casella** [comment]
The commit that broke this test case was made between Apr 24 and May 14, see [regression report](https://libraries.openmodelica.org/branches/history/master/2022-04-24%2011:23:12..2022-05-14%2000:36:35.html).

@mahge, would you mind having a look?

**Comment by perost** [comment]
I get some earlier error messages from the backend:
```
function BackendEquation.solveEquation failed w.r.t jointRRR1.n_b[3]
ARRAY_EQUATION: jointRRR1.n_bCREF jointRRR1 [model Modelica.Mechanics.MultiBody.Joints.Assemblies.JointRRR] .n_b [Real[3]] CREFTYPE:Real[3]

===================
function BackendEquation.solveEquation failed w.r.t jointRRR1.n_b[2]
ARRAY_EQUATION: jointRRR1.n_bCREF jointRRR1 [model Modelica.Mechanics.MultiBody.Joints.Assemblies.JointRRR] .n_b [Real[3]] CREFTYPE:Real[3]

===================
```
These messages do not show up in the error logs from the library testing, possibly because they are not proper error messages and just regular print statements for some reason.

**Comment by casella** [comment]
Aha, I didn't notice that. I guess this is on @kabdelhak then.

@perost, do you have some time to try some bisection and narrow down the search of the offending commit?

**Comment by perost** [comment]
The issue is #8933, so the same commit that's responsible for most of the other regressions.

The problem seems to be that the `BackendEquation.solveEquation` gets e.g. the array equation `jointRRR1.n_b = some array expression` and then fails to solve for `jointRRR1.n_b[3]`. I tried adding a case to `ExpressionSolve.solveSimple` to handle that case:
```Modelica
  // If the lhs isn't subscripted but the variable to solve is,
  // check if they reference the same variable and subscript the rhs if they are.
  case (DAE.CREF(componentRef = cr1),_,DAE.CREF(componentRef = cr))
    guard listEmpty(ComponentReference.crefLastSubs(cr1)) and
          ComponentReference.crefEqual(ComponentReference.crefStripLastSubs(cr), cr1) and
          not Expression.expHasCrefNoPreOrStart(inExp2, cr1)
    then
      (Expression.applyExpSubscripts(inExp2, ComponentReference.crefLastSubs(cr)), {});
```
That fixed the error messages from `BackendEquation.solveEquation`, but we still get warnings from the initialization and the generated code is still wrong. So it seems the actual issue is more complicated and will probably need @kabdelhak to have a look.

**Comment by casella** [comment]
After #8981 PlanarLoops_analytic [runs again](https://libraries.openmodelica.org/branches/master/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Mechanics.MultiBody.Examples.Loops.PlanarLoops_analytic.sim) and [verifies correctly](https://libraries.openmodelica.org/branches/master/Modelica_4.0.0/Modelica_4.0.0.html). 

We need to check that the values of the shape vectors are now fine, but that is already covered by #8938.
